src/ppx/window_android.cpp

```diff
@@ -27,18 +31,114 @@ namespace ppx {
 class WindowImplAndroid : public Window
 {
 public:
-    using Window::Window;
-    void FillSurfaceInfo(grfx::SurfaceCreateInfo* pCreateInfo) const final
-    {
-        pCreateInfo->androidAppContext = App()->GetAndroidContext();
-    }
+    WindowImplAndroid(Application*);
+
+    Result Create(const char* title) final;
+    bool   IsRunning() const final;
+    void   ProcessEvent() final;
+
+    WindowSize Size() const final;
+
+    void FillSurfaceInfo(grfx::SurfaceCreateInfo*) const final;
+
+    // Android callbacks.
+    void    OnAppCmd(int32_t cmd);
+    int32_t OnInputEvent(AInputEvent*);
+
+private:
+    android_app* mAndroidApp  = nullptr;
+    bool         mWindowReady = false;
+    WindowSize   mSize        = {};
 };
 
 std::unique_ptr<Window> Window::GetImplAndroid(Application* pApp)
 {
     return std::unique_ptr<Window>(new WindowImplAndroid(pApp));
 }
 
+extern "C" void WindowImplAndroidOnAppCmd(android_app* pApp, int32_t cmd)
+{
+    static_cast<WindowImplAndroid*>(pApp->userData)->OnAppCmd(cmd);
+}
+
+extern "C" int32_t WindowImplAndroidOnInputEvent(android_app* pApp, AInputEvent* event)
+{
+    return static_cast<WindowImplAndroid*>(pApp->userData)->OnInputEvent(event);
+}
+
+WindowImplAndroid::WindowImplAndroid(Application* pApp)
+    : Window(pApp), mAndroidApp(pApp->GetAndroidContext())
+{
+    mAndroidApp->onAppCmd     = WindowImplAndroidOnAppCmd;
+    mAndroidApp->onInputEvent = WindowImplAndroidOnInputEvent;
+    mAndroidApp->userData     = this;
+}
+
+Result WindowImplAndroid::Create(const char* title)
```

Comment:
nit pTitle?

Reply:
That's a C style string, pTitle sounds like const char**